### PR TITLE
sql: add ALTER RANGE RELOCATE

### PR DESCRIPTION
**Pull request by lunevalex**
This commit introduces a new ALTER RANGE RELOCATE
command, which will allow an admin to move a lease
or replica for a specific range. Unlike ALTER TABLE RELOCATE
this command works on a range_id, which makes it a lot easier
to use since the user does not have to worry about range keys, which
are difficult to deal with in an emergncy situation.

Release note (sql change): Introduce new SQL syntax ALTER RANGE RELOCATE
to move a lease or replica between stores. This is helpful in an emergency
situation to relocate data in the cluster.

### CODE REVIEW

File: pkg/sql/parser/sql.y
Comment (otan) [inline pkg/sql/parser/sql.y:847]: nit: put it in alphabetical order (not sure how EXPERIMENTAL_RELOCATE ended up here)

File: pkg/sql/parser/sql.y
Comment (otan) [inline pkg/sql/parser/sql.y:1994]: nit: indentation

File: pkg/sql/parser/sql.y
Comment (otan) [inline pkg/sql/parser/sql.y:844]: you'll probably want to add some parsing tests in `pkg/sql/parser/testdata/` (`make test PKG=./pkg/sql/parser TESTS=TestParse TESTFLAGS='-rewrite'` to rewrite)

**Comment by lunevalex** [comment]
@tbg @otan this is ready for review.

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 12 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72305)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lunevalex, @nvanbenschoten, @otan, and @tbg)

___
*[pkg/sql/parser/sql.y, line 1994 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MnSeqaUE19tyHauY0pl-r1-1994:-MosIQ8_BEtjlf17OJ0O:bbp9ile) ([raw file](https://github.com/cockroachdb/cockroach/blob/9843c0dbae0490cc497faa4462aa0b9795707070/pkg/sql/parser/sql.y#L1994)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: indentation
</blockquote></details>

specifically, use spaces not tabs
___
*[pkg/sql/parser/sql.y, line 1981 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MosIFbP5cVNm39Fd1xT:-MosIFbQ8mQ7V4B3KD92:bqrqt3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ee93af61e7208769eecfcfee951434d1aa8e723/pkg/sql/parser/sql.y#L1981)):*
> ```Yacc
> 
> alter_range_relocate_lease_stmt:
>   ALTER RANGE relocate_kw LEASE TO iconst64 FOR select_stmt
> ```

Can you either extend that iconst64 to support expressions and/or placeholders, or file a followup issue to do just this.

In particular the placeholders will be important for tests.

**Review by lunevalex — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72305)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @knz, @nvanbenschoten, @otan, and @tbg)

___
*[pkg/sql/parser/sql.y, line 844 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MnSeqaUE19tyHauY0pl-r1-844:-Mosi4mfBEItScpiJCiK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9843c0dbae0490cc497faa4462aa0b9795707070/pkg/sql/parser/sql.y#L844)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

you'll probably want to add some parsing tests in `pkg/sql/parser/testdata/` (`make test PKG=./pkg/sql/parser TESTS=TestParse TESTFLAGS='-rewrite'` to rewrite)
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 1981 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MosIFbP5cVNm39Fd1xT:-MoskcN22CzlKtJ2rkoG:b-axib1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ee93af61e7208769eecfcfee951434d1aa8e723/pkg/sql/parser/sql.y#L1981)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you either extend that iconst64 to support expressions and/or placeholders, or file a followup issue to do just this.

In particular the placeholders will be important for tests.
</blockquote></details>

I will add a follow up once this lands.

File: pkg/kv/kvserver/client_lease_test.go
Comment (tbg) [inline pkg/kv/kvserver/client_lease_test.go:984]: Wouldn't be surprised if this test ended up flaky, mind giving this a bit of extra love before merge? I know CI stresses it but sometimes not enough.

File: pkg/sql/opt/optbuilder/alter_range.go
Comment (tbg) [inline pkg/sql/opt/optbuilder/alter_range.go:30]: I don't know what this is about, but it might be worth adding a comment.

**Review by tbg — APPROVED** [review]
LGTM from kv. I assume you'll be waiting for an LGTM on the syntax plumbing as well, since I am not a good reviewer for those.

**Review by lunevalex — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72305)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @knz, @lunevalex, @nvanbenschoten, @otan, and @tbg)

___
*[pkg/kv/kvserver/client_lease_test.go, line 984 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MoorLneYNW7MYgRlM07-r4-984:-MotGmqYARqG3jPldgmC:bgvs52k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b73251391c9848966413a7503fa7a50bd43e158/pkg/kv/kvserver/client_lease_test.go#L984)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Wouldn't be surprised if this test ended up flaky, mind giving this a bit of extra love before merge? I know CI stresses it but sometimes not enough.
</blockquote></details>

will do 
___
*[pkg/sql/opt/optbuilder/alter_range.go, line 30 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MnSeqaUE19tyHauY0pf-r4-30:-MotGtfoCeHBGuxBaqIT:bb67cvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b73251391c9848966413a7503fa7a50bd43e158/pkg/sql/opt/optbuilder/alter_range.go#L30)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I don't know what this is about, but it might be worth adding a comment.
</blockquote></details>

 ack

**Review by arulajmani — COMMENTED** [review]
This is pretty cool! :100: 

Only a few nits from me, otherwise :lgtm: 

> Reviewed 4 of 29 files at r1, 18 of 30 files at r2, 11 of 12 files at r3, 2 of 2 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-:-MouX4fF1piSLsNggxJW:bha6hev)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @knz, @lunevalex, @nvanbenschoten, @otan, and @tbg)

___
*[pkg/sql/relocate_range.go, line 63 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-Mou88kY-NSvcYvopAkd:-Mou88kZDP-aZqqRsQ6f:bqi83c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L63)):*
> ```Go
> 	}
> 	if !n.relocateLease {
> 		if n.fromStoreID <= 0 {
> ```

nit: you could collapse these two conditionals into a single one
___
*[pkg/sql/relocate_range.go, line 82 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-Motu5IL46PH806qtBVS:-Motu5IM5-_luQdD3lnN:b-prc29h) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L82)):*
> ```Go
> 	return nil
> }
> func (n *relocateRange) Next(params runParams) (bool, error) {
> ```

nit: new line above
___
*[pkg/sql/relocate_range.go, line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MouAFoy2l_Ye87LVy7a:-MouAFoy2l_Ye87LVy7b:b-fw47zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L130)):*
> ```Go
> 
> func relocate(params runParams, req relocateRequest) (*roachpb.RangeDescriptor, error) {
> 	rangeDesc, err := lookupRangeDescriptorByRangeID(params.ctx, params.extendedEvalCtx.ExecCfg.DB, req.rangeID)
> ```

nit: wrap line at 100 chars (this and a couple more below)
___
*[pkg/sql/relocate_range.go, line 136 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MouAxFaAB-hbjQH4SbN:-MouAxFaAB-hbjQH4SbO:b-jvkqg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L136)):*
> ```Go
> 
> 	if req.relocateLease {
> 		if err := params.p.ExecCfg().DB.AdminTransferLease(params.ctx, rangeDesc.StartKey, req.toStoreDesc.StoreID); err != nil {
> ```

nit: We could simply return without checking if the error is nil here. 
___
*[pkg/sql/relocate_range.go, line 139 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MouB7VoDTE10qSsAAYb:-MouB7VoDTE10qSsAAYc:b1zdf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L139)):*
> ```Go
> 			return rangeDesc, err
> 		}
> 	} else {
> ```

nit: If you accept the suggestion above you could inline everything in this else block and buy a bit more indentation. 
___
*[pkg/sql/relocate_range.go, line 180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MouNctl4-X7jziGoxaW:-MouNctm8cX6lMt-aoV0:b1jme2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L180)):*
> ```Go
> 					}
> 					// In small enough clusters it's possible for the same range
> 					// descriptor to be stored in both meta1 and meta2. This
> ```

This is probably because I'm unfamiliar with the area, but could you explain the effect of a range descriptor being stored in both meta1 and meta2 for my edification?

**Comment by lunevalex** [comment]
___
*[pkg/sql/relocate_range.go, line 180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MouNctl4-X7jziGoxaW:-Mp8YFZx-cg0L90aNknJ:bl80pmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L180)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This is probably because I'm unfamiliar with the area, but could you explain the effect of a range descriptor being stored in both meta1 and meta2 for my edification?
</blockquote></details>

This has to do with https://github.com/cockroachdb/cockroach/blob/35109acda3a49d14dadbd66e562c9da5e38836e6/pkg/kv/kvserver/addressing.go#L92. We store the same range descriptor in meta1 and meta2 in cases where the range spans the meta and data keyspaces so that a meta1 lookup is always sufficient to find all meta2 entries and a meta2 lookup is always sufficient to find all data entries. As pointed out by Nathan on an internal chat thread.

**Comment by lunevalex** [comment]
TFTR!
bors r=tbg,arulajmani

**Review by arulajmani — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-:-MpCsP-LC78A2U-orHzk:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @knz, @lunevalex, @nvanbenschoten, @otan, and @tbg)

___
*[pkg/sql/relocate_range.go, line 180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MouNctl4-X7jziGoxaW:-MpCrsqUAfbO-kKyWbbI:boupq8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L180)):*
<details><summary><i>Previously, lunevalex wrote…</i></summary><blockquote>

This has to do with https://github.com/cockroachdb/cockroach/blob/35109acda3a49d14dadbd66e562c9da5e38836e6/pkg/kv/kvserver/addressing.go#L92. We store the same range descriptor in meta1 and meta2 in cases where the range spans the meta and data keyspaces so that a meta1 lookup is always sufficient to find all meta2 entries and a meta2 lookup is always sufficient to find all data entries. As pointed out by Nathan on an internal chat thread. 
</blockquote></details>

Looks like we're iterating over both meta1 and meta2 here, and returning the first range descriptor entry that we find. Maybe we should add a bit to this comment to say that even if two copies of the range descriptor are present, we don't care, and simply return the first one we find?

**Review by lunevalex — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72305)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @knz, @lunevalex, @nvanbenschoten, @otan, and @tbg)

___
*[pkg/sql/relocate_range.go, line 180 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72305#-MouNctl4-X7jziGoxaW:-MpD2m6JCXhz-SOVl--R:bb7oryy) ([raw file](https://github.com/cockroachdb/cockroach/blob/849ae3b524b54493d472bd70122e2b4a8c2444b9/pkg/sql/relocate_range.go#L180)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Looks like we're iterating over both meta1 and meta2 here, and returning the first range descriptor entry that we find. Maybe we should add a bit to this comment to say that even if two copies of the range descriptor are present, we don't care, and simply return the first one we find?
</blockquote></details>

sure

**Comment by lunevalex** [comment]
blathers backport 21.2

**Pull request merged**
